File: app/src/main/java/com/github/shadowsocks/preferences/SummaryEditTextPreference.java
package com.github.shadowsocks.preferences;

import android.content.Context;
import android.preference.EditTextPreference;
import android.text.TextUtils;
import android.util.AttributeSet;

public class SummaryEditTextPreference extends EditTextPreference {

    private final CharSequence mDefaultSummary;

    public SummaryEditTextPreference(Context context, AttributeSet attrs) {
        this(context, attrs, android.R.attr.editTextPreferenceStyle);
    }

    public SummaryEditTextPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mDefaultSummary = getSummary();
    }

    @Override
    public void setText(String text) {
        super.setText(text);
        setSummary(text);
    }

    @Override
    public void setSummary(CharSequence summary) {
        if (TextUtils.isEmpty(summary)) {
            super.setSummary(mDefaultSummary);
        } else {
            super.setSummary(summary);
        }
    }
}

File: app/src/main/java/com/bige0/shadowsocksr/preferences/SummaryEditTextPreference.kt
package com.bige0.shadowsocksr.preferences

import android.content.*
import android.preference.*
import android.util.*

class SummaryEditTextPreference @JvmOverloads constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int = android.R.attr.editTextPreferenceStyle) : EditTextPreference(context, attrs, defStyleAttr)
{
	private val mDefaultSummary: CharSequence? = summary

	override fun setText(text: String)
	{
		super.setText(text)
		summary = text
	}

	override fun setSummary(summary: CharSequence)
	{
		if (summary.isEmpty())
		{
			super.setSummary(mDefaultSummary)
		}
		else
		{
			super.setSummary(summary)
		}
	}
}
